File: matlab-libtiepie-master/Reference MATLAB Files/FiberCharacterisationJUG.m
clear all;
close all;

%% Saving name
name = 'DM-0001-DC1_on_top ';

%% Settings ref perfect mirror
refCh = [2.340 1.123]; % Signal with perfect mirror in V [Ch1 Ch2]

%% Settings for oscilloscope
Lrec = 10000; % Recording length
Fs = 1e6; % Sampling freq

ampRange = 4; % oscillo amplitude in V

%%
global h; % make h a global variable so it can be used outside the main
          % function. Useful when you do event handling and sequential move
%% Create Matlab Figure Container
fpos    = get(0,'DefaultFigurePosition'); % figure default position
fpos(3) = 650; % figure window size;Width
fpos(4) = 450; % Height
f = figure('Position', fpos,...
           'Menu','None',...
           'Name','APT GUI');
%% Create ActiveX Controller
%h = actxcontrol('MG17Piezo1.APTPiezoCtrl.1',[20 20 600 400 ], f);
h = actxcontrol('MG17SYSTEM.MG17SystemCtrl.1', [20 20 600 400 ], f);

%% Initialize
% Start Control
% Set the Serial Number
SN = 81837345; % put in the serial number of the hardware
h_Piezo = actxcontrol('MGPIEZO.MGPiezoCtrl.1', [20 20 600 400 ], f);

h_Piezo.StartCtrl;

set(h_Piezo,'HWSerialNum', SN);
% Indentify the device

h_Piezo.Identify;
pause(2); % waiting for the GUI to load up;
%h_Piezo.SetVoltageOutputLimit
%h_Piezo.SetVoltOutput

%%
h_Piezo.SetVoltOutput(0,20)

%% TiePie stuffs
system('taskkill /im MultiChannel.exe'); % Close TiePie in case it was open

if verLessThan('matlab', '8')
    error('Matlab 8.0 (R2012b) or higher is required.');
end

% Open LibTiePie and display library info if not yet opened:
import LibTiePie.Const.*
import LibTiePie.Enum.*

if ~exist('LibTiePie', 'var')
    % Open LibTiePie:
    LibTiePie = LibTiePie.Library
end

% Enable network search:       
LibTiePie.Network.AutoDetectEnabled = true;

% Search for devices:
LibTiePie.DeviceList.update();

Vvolt = []; % Volt vector
Vch = []; % Channel signal vector

figure('Name','Live plot')

%% Set Acquire
samplerate = 1;
bufferdelay = 0.1;
df=(1.25*10^6); 
    %samplerate = 1024;
    %bufferdelay = 134.218*10^-3;
    %df=(12.2*10^4);
    %samplerate = 8192;
    %bufferdelay = 1.074;ju
    %df=(1.52*10^4);
dt = 1/df;

clear total;
iitotal=20;
for tt=1%:25
trigger=0;
voltstop=0;
hystTest = [3:1:65]
for ii = hystTest % iitotal:-0.25:3
volt=(ii-1);
%for ii = 70:-1:1

if volt>75
    volt = 75;
end
%if voltstop>0
%    volt=voltstop;
%end
h_Piezo.SetVoltOutput(0,volt)

pause(.01)
Vvolt = [Vvolt; volt];

%% tiepie in the loop
% Try to open an oscilloscope with block measurement support and a generator in the same device:
    clear scp;
    clear gen;
    for k = 0 : LibTiePie.DeviceList.Count - 1
        item = LibTiePie.DeviceList.getItemByIndex(k);
        if item.canOpen(DEVICETYPE.OSCILLOSCOPE) && item.canOpen(DEVICETYPE.GENERATOR)
            scp = item.openOscilloscope();
            if ismember(MM.BLOCK, scp.MeasureModes)
                gen = item.openGenerator();
                break;
            else
                clear scp;
            end
        end
    end
    clear item



    if exist('scp', 'var') && exist('gen', 'var')
        % Oscilloscope settings:

        % Set measure mode:
        scp.MeasureMode = MM.BLOCK;

        % Set sample frequency:
        scp.SampleFrequency = Fs; % 1 MHz

        % Set record length:
        scp.RecordLength = Lrec; % 10000 Samples

        % Set pre sample ratio:
        scp.PreSampleRatio = 0; % 0 %

        % For all channels:
        for ch = scp.Channels
            % Enable channel to measure it:
            ch.Enabled = true;

            % Set range:
            ch.Range = ampRange; 

            % Set coupling:
            ch.Coupling = CK.DCV; % DC Volt

            % Release reference:
            clear ch;
        end

        % Set trigger timeout:
        scp.TriggerTimeOut = 1; % 1 s

        % Disable all channel trigger sources:
        for ch = scp.Channels
            ch.Trigger.Enabled = false;
            clear ch;
        end

        % Locate trigger input:
        triggerInput = scp.getTriggerInputById(TIID.GENERATOR_NEW_PERIOD); % or TIID.GENERATOR_START or TIID.GENERATOR_STOP

        if triggerInput == false
            clear triggerInput;
            clear scp;
            clear gen;
            error('Unknown trigger input!');
        end

        % Enable trigger input:
        triggerInput.Enabled = true;

        % Release reference to trigger input:
        clear triggerInput;

        % Generator settings:

        % Set signal type:
        gen.SignalType = ST.SINE;

        % Set frequency:
        gen.Frequency = 1e3; 

        % Set amplitude:
        gen.Amplitude = 2; % 2 V

        % Set offset:
        gen.Offset = 0; % 0 V

        % Enable output:
        gen.OutputOn = true;

        % Print oscilloscope info:
        %display(scp);

        % Print generator info:
        %display(gen);

        % Start measurement:
        scp.start();

        % Start signal generation:
        gen.start();

        % Wait for measurement to complete:
        while ~scp.IsDataReady
            pause(10e-3) % 10 ms delay, to save CPU time.
        end

        % Stop generator:
        gen.stop();

        % Disable output:
        gen.OutputOn = false;

        % Get data:
        arData = scp.getData();
        Vch = [Vch; mean(arData)];

        % Get all channel data value ranges (which are compensated for probe gain/offset):
        clear darRangeMin;
        clear darRangeMax;
        for i = 1 : length(scp.Channels)
            [darRangeMin(i), darRangeMax(i)] = scp.Channels(i).getDataValueRange();
        end

%         % Plot results:
%         if mod(incre,40) == 1
%             figure
%             plot((1:scp.RecordLength) / scp.SampleFrequency, arData);
%             xlabel('Time [s]');
%             ylabel('Amplitude [V]');
%             legend('Channel 1','Channel 2');
%             grid on;
%         end

        % Close oscilloscope:
        clear scp;

        % Close generator:
        clear gen;
    else
        error('No oscilloscope available with block measurement support or generator available in the same unit!');
    end
    
    plot(Vvolt,Vch*1e3);
    grid on;
    xlabel('Input piezo voltage (to be converted in length) [V]');
    ylabel('Output voltage [mV]');
    legend('Channel 1','Channel 2')

end
end

h_Piezo.SetVoltOutput(0,20)

clear LibTiePie % necessary to use TiePie after MatLab

system('cd C:\Users\julie\OneDrive\Documents\MATLAB & start MatLab.tps'); % Reopen TiePie

%% Plot 
close all 

dist = Vvolt*0.32; % convert in micrometers

subplot(2,2,1)
title('Percentage of Ref')
plot(dist,Vch./refCh*100);
grid on;
xlabel('Distance from mirror (arbitrary 0) [\mum]');
ylabel('Coupling [%]');
legend('Channel 1','Channel 2')


subplot(2,2,2)
title('Real voltage value')
plot(Vvolt,Vch*1e3);
grid on;
xlabel('Input piezo voltage (to be converted in length) [V]');
ylabel('Output voltage [mV]');
legend('Channel 1','Channel 2')

max(Vch(:,1)./refCh(1)*100)

% figure
% dist = Vvolt*25/75; % Distance en micron
% plot(dist,Vch*1e3);
% grid on;
% xlabel('Distance from the starting point (close to the mirror) [\mum]');
% ylabel('Output voltage [mV]');
% legend('Channel 1','Channel 2')

subplot(2,2,3)
title('Gradient with distance')
plot(dist,gradient(Vch(:,1)/refCh(1)*100,25/75));
grid on; hold on;
plot(dist,gradient(Vch(:,2)/refCh(2)*100,25/75));
xlabel('Distance from mirror (arbitrary 0) [\mum]');
ylabel('Sensitivity [%/\mum]');
legend('Channel 1','Channel 2')


subplot(2,2,4)
title('voltage values')
plot(dist,Vch*1e3);
grid on;
xlabel('Distance from mirror (arbitrary 0) [\mum]');
ylabel('Output voltage [mV]');
legend('Channel 1','Channel 2')

saveas(gcf,[name,' - Coupling - ',datestr(now, 'dd-mmm-yyyy HH-MM-SS'),'.png']);

% figure
% plot(dist,-gradient(Vch(:,2)/0.227*70,25/75));
% grid on;
% xlabel('Distance from the starting point (close to the mirror) [\mum]');
% ylabel('Output voltage [mV]');
% legend('Channel 1','Channel 2')

%% Saving variables
save([name,' - Fiber char - ',datestr(now, 'dd-mmm-yyyy HH-MM-SS')]);

return 

File: python-libtiepie-master/Python Routines/FiberCharacterisationJUG.py
# -*- coding: utf-8 -*-
"""
Created on Sunday 28.11.2021 at 14:09

Miraex - Fiber characterization routine

@author: Claudio Jaramillo
"""

from __future__ import print_function
import libtiepie
import time
import matplotlib.pyplot as plt
import os
import sys

import numpy as np
import MiraexLib.plot as miraex_plt
import MiraexLib.misc as miraex_misc
import MiraexLib.analysis as miraex_anls
from MiraexLib.printinfo import*

from cubini.KPZ101 import KPZ101

"""
IMPORTANT DOCUMENTATION :

This script depends on the [Cubini library](https://github.com/Schlabonski/cubini).

To install follow the steps in the READ ME of the repo.

If the line :

    python3 setup.py install

fails, try using :

    python setup.py install

For this script to work properly, the libraries :
 - PyUSB
 - LibUSB

Need to be installed.

Install them both first by using :

    conda install -c conda-forge pyusb
    conda install -c conda-forge libusb

from the anaconda cmd line after having activated the relevant environment.

> Environment Miraex_2021_12_01 is the earliest to have this dependency installed.

Then, go to : https://github.com/libusb/libusb/releases/tag/v1.0.24

Download the zip file, and open it.

Unzip using 7zip into a temp dir.

If on 64-bit Windows, copy

    MinGW64\dll\libusb-1.0.dll

into

    C:\windows\system32

If on 32-bit windows, copy

    MinGW32\dll\libusb-1.0.dll

into

    C:\windows\SysWOW64.

This should solve the ''No backend available'' exepction raised at init of the cuboids.

Warning : use the set_max_voltage method before the set_output_voltage method to avoid crashes.
"""

# %% Generic parameters

name = 'DM-0001-DC1_on_top '

# %% Cuboid parameters

V_max = 75  # 75 V maximum voltage
V_default = 20  # V default voltage

# %% Settings ref perfect mirror
refCh = [2.340, 1.123]  # Signal with perfect mirror in V [Ch1 Ch2], reference

# %% Settings for oscilloscope
Lrec = 10000  # Recording length
Fs = 1e6  # Sampling freq

ampRange = 4  # Oscilloscope amplitude in V

# make h a global variable so it can be used outside the main function. Useful when you do event handling and sequential move
global h

# %% Initialization

# list of serial numbers
cuboids = [29250837]

# connect to all modules
cubinis = []

print('here')

for sn in cuboids:
    # try:
    cubini = KPZ101(serial_number=sn)
    time.sleep(0.1)
    cubini.set_input_mode()
    cubinis.append(cubini)
    print('KPZ101 {} found.'.format(sn))

"""
    except Exception as e:
        print(e)
        print('KPZ101 {} not found.'.format(sn))
"""
"""
# sequentially run a test sequence for each cube
for kpz in cubinis:
    kpz.set_max_voltage(V_max)
    kpz.enable_output()
    kpz.set_output_voltage(0)
    time.sleep(0.5)
    for v in np.linspace(0, 50, 25):
        kpz.set_output_voltage(v)
        time.sleep(0.1)
    kpz.set_output_voltage(0)
    kpz.disable_output()
"""


# set initial parameters sequentially
for kpz in cubinis:
    kpz.set_max_voltage(V_max)
    kpz.enable_output()
    kpz.set_output_voltage(V_default)


# %% TiePie initialization

# Print library info:
print_library_info()

# Enable network search:
libtiepie.network.auto_detect_enabled = True

# Search for devices:
libtiepie.device_list.update()

# %% Parameter set before code

Vvolt = []  # Volt vector

average_list = []
Vect_AmpRMS = []
data_no_offset = []

# Set acquire
samplerate = 1
bufferdelay = 0.1
df = (1.25*(10**6))

dt = 1/df

iitotal = 20

# %% Main loop

for ii in range(1, 1, 1):  # start, stop, step
    trigger = 0
    voltstop = 0
    hystTest = np.arange(3, 65, 1)

    for tt in hystTest:
        print('hello world')

        # Try to open an oscilloscope with block measurement support and a generator in the same device:
        scp = None
        gen = None
        for item in libtiepie.device_list:
            if (item.can_open(libtiepie.DEVICETYPE_OSCILLOSCOPE)) and (item.can_open(libtiepie.DEVICETYPE_GENERATOR)):
                scp = item.open_oscilloscope()
                if scp.measure_modes & libtiepie.MM_BLOCK:
                    gen = item.open_generator()
                    break
                else:
                    scp = None

        if scp and gen:
            try:
                # Oscilloscope settings:

                # Set measure mode:
                scp.measure_mode = libtiepie.MM_BLOCK

                # Set sample frequency:
                scp.sample_frequency = Fs  # Hz

                # Set record length:
                scp.record_length = Lrec  # nb of samples

                # Set pre sample ratio:
                scp.pre_sample_ratio = 0  # 0 %

                # For all channels:
                for ch in scp.channels:
                    # Enable channel to measure it:
                    ch.enabled = True

                    # Set range:
                    ch.range = ampRange  # V

                    # Set coupling:
                    ch.coupling = libtiepie.CK_DCV  # DC Volt

                # Set trigger timeout:
                scp.trigger_time_out = 1  # 1 s

                # Disable all channel trigger sources:
                for ch in scp.channels:
                    ch.trigger.enabled = False

                # Locate trigger input:
                # or TIID_GENERATOR_START or TIID_GENERATOR_STOP
                trigger_input = scp.trigger_inputs.get_by_id(
                    libtiepie.TIID_GENERATOR_NEW_PERIOD)

                if trigger_input is None:
                    raise Exception('Unknown trigger input!')

                # Enable trigger input:
                trigger_input.enabled = True

                # Generator settings:

                # Set signal type:
                gen.signal_type = libtiepie.ST_SINE

                # Set frequency:
                gen.frequency = 1e3  # 1 kHz

                # Set amplitude:
                gen.amplitude = 2  # 2 V

                # Set offset:
                gen.offset = 0  # 0 V

                # Enable output:
                gen.output_on = True

                # Print oscilloscope info:
                print_device_info(scp)

                # Print generator info:
                print_device_info(gen)

                # Start measurement:
                scp.start()

                # Start signal generation:
                gen.start()

                # Wait for measurement to complete:
                while not scp.is_data_ready:
                    time.sleep(0.01)  # 10 ms delay, to save CPU time

                # Stop generator:
                gen.stop()

                # Disable output:
                gen.output_on = False

                # Get data:
                data = scp.get_data()
                np_data = np.array(data)  # numpy array
                #print(f'{np_data.shape = }')

                # average value of the data fetched during 1 loop per channel
                # The first index is the iteration index
                # The second index is the avg for channel index
                average_list.append([np.mean(np_data[i])
                                    for i in range(len(np_data))])

                # Substract the mean of np_data to each element of np_data during 1 loop
                '''
                Processed data will have n number of elements, each has nb_of_channels sub-dimensions that represent the nb of channels that we are working with in the oscilloscope.

                The number of elements n is the number of iterations of the loop. It increases as the test time increases.
                '''
                # By substracting the mean we remove the DC offset for each channel
                data_no_offset.append([np_data[i] - average_list[i]
                                      for i in range(len(np_data))])

                del np_data

                # Get all channel data value ranges (which are compensated for probe gain/offset)
                data_range_min_max = []
                for channel in scp.channels:
                    data_range_min_max.append(
                        [channel._get_data_value_min(), channel._get_data_value_max()])

                # Creation of lists to plot
                # RMS of the data fetched in 1 loop iteration, for each channel
                AmpRMS = [np.sqrt(np.mean((np.array(data_no_offset[i]))**2))
                          for i in range(len(data_no_offset))]

                # Create a vector with all the RMS values of all loops, updates each loop
                Vect_AmpRMS.append(AmpRMS)
                Vect_AmpRMS_np = np.array(Vect_AmpRMS)

                # Create a Channel first array of RMS values for plotting and better addressing of the data
                Channel_AmpRMS = np.transpose(Vect_AmpRMS_np)

                '''
                # Output CSV data:
                csv_file = open('OscilloscopeGeneratorTrigger.csv', 'w')
                try:
                    csv_file.write('Sample')
                    for i in range(len(data)):
                        csv_file.write(';Ch' + str(i + 1))
                    csv_file.write(os.linesep)
                    for i in range(len(data[0])):
                        csv_file.write(str(i))
                        for j in range(len(data)):
                            csv_file.write(';' + str(data[j][i]))
                        csv_file.write(os.linesep)

                    print()
                    print('Data written to: ' + csv_file.name)


                finally:
                    csv_file.close()
                '''

            except Exception as e:
                print('Exception: ' + e.message)
                sys.exit(1)

            # Close oscilloscope:
            del scp

            # Close generator:
            del gen

        else:
            print('No oscilloscope available with block measurement support or generator available in the same unit!')

        for i in range(len(average_list)):
            miraex_plt.DynamicPlot2(Vvolt, average_list[i],
                                    'Duration [s]', 'RMS', 'Test')


# Reset cuboids to default parameters
for kpz in cubinis:
    kpz.set_max_voltage(V_max)
    kpz.enable_output()
    kpz.set_output_voltage(V_default)
